src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningKeyMapper.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.amazonaws.services.dynamodbv2.model.ScalarAttributeType;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.primitives.UnsignedBytes;
+import com.salesforce.dynamodbv2.mt.context.MtAmazonDynamoDbContextProvider;
+import com.salesforce.dynamodbv2.mt.mappers.metadata.PrimaryKey;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Objects;
+
+class HashPartitioningKeyMapper {
+
+    @VisibleForTesting
+    static final String DELIMITER = ".";
+
+    private static final String DELIMITER_PATTERN = "\\.";
+
+    static class HashPartitioningKeyPrefixFunction {
+
+        static AttributeValue toPhysicalHashKey(String context, String virtualTableName, int bucket) {
+            String stringValue = String.join(DELIMITER, context, virtualTableName, String.valueOf(bucket));
```

Comment:
I wonder if it would make sense to serialize these values using an actual binary encoding, i.e., encode bucket as four byte number (or maybe 2 bytes, depending on how big we think the max will get) and context and virtual table name as variable-length strings (delimited & escaped). That would make it more space efficient, especially, if we allow different context types in the future (e.g., int).

If we stick with a string, we could probably store it as such to make it easier to work with the system (since dynamodb internally just converts the string into a byte array the same way).

Reply:
Yeah that makes sense. So for example, some format like bucket + contextBytesLength + contextBytes + virtualTableNameBytes?
It can also make it easier to figure out only the bucket number given a physical hash key, though I'm not sure if we need that.

Reply:
We could use something like the ZOS encoding:
https://git.soma.salesforce.com/salesforce-zero/zero-object-service/blob/a5d7a2203b074e09fc6659c3eeb567bbef30eec3/zos/src/main/java/com/salesforce/zos/dynamo/processors/Codec.java

I wasn't necessarily proposing to change the order in which we store the three fields, though I suppose we could do that (not sure it makes a difference).

So far, we've been restricting the type (String) and characters for context and virtual table name, so that we can keep a consistent interface with `byTable`. Which is why I think using `.` as the separator works. However, if we move away from supporting both implementations and introduce explicit multi-tenant tables, we may would want to reconsider that, i.e., allow using different context types, such as integer, or other characters. In that case, we'd need some kind of escape/separator or length approach.

Reply:
Cool I see. Since we're already storing composite keys in the physical RK using the length approach, and the lengths of context and table can be stored in shorts (can't be more than 2048), maybe we can also use the length approach here? (I find it simpler than escaping.)

Reply:
Sounds good